Rewrite data flow in app to remove races and sync bugs

dmsnell opened the pull request:
## Description

We've been unable to resolve the worst sync bugs in the application and feature development has been artificially limited by the fact that we have to fight a number of data-flow issues tightly coupling different levels of the application's architecture together.

In this branch, we're ripping apart the entire app state and Simperium data flow to rebuild it in order to remove a number of those couplings and races.

## Blockers

 - [ ] ~Enter key not working in Firefox, programmatic edits to note content not working - execCommand~
 - [x] Cannot quit app if unsync'd notes exist (for example, emptied trash, notes in bucket, missing from ghost, looping to sync notes) [`beforeunload` works differently in Electron]
 - [ ] Deleted-forever notes may not actually delete (refresh the app and they might come back). This may be related to the items below on adding a *delete-queue*
 - [x] Add cut/copy/paste to context menu when editing #2280 
 - [x] Make sure notifications do not spam users on login
 - [x] Make sure notifications does not spam users when requesting permission
 - [x] Make sure notifications off by default
 - [ ] Add Electron confirm close dialog
 - [x] Test all keyboard shortcuts [#](https://github.com/Automattic/simplenote-electron/pull/2148#issuecomment-675743700)
  - [x] Fix broken shortcuts (Done except for "next match?") #2284 
 - [x] Debounce `editor_note_edited` analytics to follow Simperium's edits, not the editor's (@dmsnell, @codebykat
 - [x] Printing in edit mode results in blank page, preview works @belcherj #2271
 - [x] Search is missing #2313 

## Not-really blockers

 - [x] Audit tag operations to make sure affected notes are synchronized (@dmsnell)
 - [x] Only select the next note during `DELETE_NOTE_FOREVER` if the opened note is the one that was deleted
 - [x] Create checkbox, mark as completed, enter down, checkbox created is marked as complete.
     - When continuing a task list the new checkboxes will now all be unchecked if automatically created
 - [x] Disable "matching parenthesis and brackets" from the editor. they are visually wrong and not "Simple" #2283 

### Safari
 - [x] Buttons in revision history and trash view do not have a proper line-height
 - [x] Line height for search highlights is incorrect.
 - [x] Text on staging is #d4d4d4 not #fff in dark mode (should be fixed by #2272)
 - [x] Tag names in edit mode have bottom chopped off, edit tag a `qqq`, observe
 - [x] In search field descenders are cut off, type `qqq`
 - [x] When coming back online sever connection does not go fully green

### Firefox

 - [ ] ~When you have the publish dialog open in Firefox and Chrome, toggle on publishing, it then toggles off~
 - [ ] ~Managed to get a note to go out sync after being offline~
 - [ ] ~Select all in context menu doesn't work~

## Possible sync issues
### Unsynchronized changes before indexing operation

While this is not present in the branch as-is I have been exploring periodically calling for an index operation to catch any notes  or edits we might have mistakenly missed from the server.

When I did this I noticed that if I had unsynchronized changes in a note they were immediately wiped away during an indexing operation.

```js
// in dev console
dispatch({type: 'CREATE_NOTE'}); // [Enter]
// grab returned `noteId` - e.g. "b5f6248c-b81e-4096-9960-e3bb73dfaaf3"

dispatch({type: 'EDIT_NOTE', noteId: "b5f6248c-b81e-4096-9960-e3bb73dfaaf3", changes: { content: 'test' }});
noteBucket.channel.sendIndexRequest(); // [Enter]
```

After running this sequence of events you can watch the `test` content wipe out during the index. It also doesn't update. I was able to get the `/old` application in a state where it saw the 'test' update but the local Electron copy reverted it. Both continued to show different data after refreshing. Once I logged-out of the Electron app and re-logged-in the update came through and it matched the content in `/old`

## Related or fixed issues

### Defects
Resolves #2171 (note display doesn't update height immediately upon change from menu bar)
Resolves #2074 (hopefully - floating IME on Korean input)
Resolves #2014 (when opened note changes so it's not in the search, it still stays open)
Resolves #1953 (when renaming tag, update in search bar if tag opened)
Resolves #1942 (maybe - can't easily select start of note content)
Resolves #1887 (renaming a tag removes it from a note)

#### Cursor position and movement
Resolves #2085 (cursor jumps to new note)
Resolves #2035 (cursor hidden at bottom of note)
Resolves #1595 (restore cursor position when flipping between edit/preview mode)
Resolves #1477 (cursor jumping in Windows)

#### Synchronization
Resolves #1938 (not restoring tags when reverting to an earlier revision)
Resolves #1641 (updates to a tag in one session don't update in other sessions until they reload)
Resolves #1640 (infinite duplication of tag name when editing tag)
Resolves #1562 (content not syncing)
Resolves #1520 (only syncing one note at a time/per session)
Resolves #1291 (quirky unsynced changes info)
Resolves #502 (data loss when editing simultaneous with iOS)
Resolves #459 (show note sync indicator)

#### Force-sync
#1897
#800

#### Ghost-Writing
Resolves #2030 (hopefully)
Resolves #1787 (hopefully)

### Enhancements
Resolves #2162 (mostly - additional syncing metadata for notes)
Resolves #1836 (remove `app-state`)
Resolves #1816 (confusing revision history ordering)
#1537 (add indicator to show syncing status)
Resolves #2036 (logging-out in one sessions logs out all sessions)
Resolves #1410 (logout is buried in app settings and hard to find)

#### Private Mode
#1924 - full functionality in Firefox private mode other than offline persistence

### Performance
Resolves #2172 (slow app)
Resolves #501 (slow loading large notes)

### Deprecations
Resolves #2117 (audit use of draftJS)
Resolves #1762 (decorator performance in draftJS)
Resolves #1026 (use of `token` in `localStorage` in `boot.js`)

### Possibly - check these
#1698 (problems with Korean input and lists at start of note)
#1619 (buggy Japanese IME conversion)
#1572 (RTL languages - checked tasks moving to the right)
#1511 (missing characters on Korean IME conversion)
#1456 (slow, possibly 4K-resolution-related)

## Status

This branch is accessible at https://staging.simplenote.com - **use at your own risk**.
The architectural rework is done and we shouldn't see any major changes before a merge.
What remains is mostly a broad bag of UI glitches and imperfections resulting from ripping
apart the old code while updating the data flow.

A few major areas of functionality are missing or incomplete, including but not limited to:
 - [x] Syncing tags
 - [x] Syncing preferences
 - [x] Showing and selecting note revisions
 - [x] Offline persistence
 - [ ] Tag Doctor to synchronize tags and fix duplicates/encoding issues
 - [x] Note Doctor to synchronize unsync'd notes
 - [ ] ~Load left-over changes from existing app - `localStorage` queue (@dmsnell)~ (leaving out for now to avoid the complexity and risk of propagating note corruption)

Scribbles from my own personal development notes:

## Discarded functionality
 - [x] Fix Safari loading
 - [x] Create tags for imported note tags
 - [x] Add markdown preview panel
 - [x] Add revisions panel
 - [x] Fix Evernote import (@belcherj)
 - [x] Add import tag/system tag
 - [x] Test text file import
 - [x] Fix exporting note id issue - using index in `notes` array vs. `noteId` (@belcherj)
 - [x] Sort notes
 - [x] Add sort index
 - [x] Add hasLoaded logic
 - [x] Reorder tags
      - [x] Tag reordering doesn't appear in Safari or Firefox #1927
 - [x] Edit tags
 - [x] Tag list does not scroll (for now we put scrolling on the parent container - see #1927) https://github.com/Automattic/simplenote-electron/pull/2210
 - [x] Reconnect settings with Electron menu
 - [x] Trash tags
 - [x] Process markdown in note list
 - [x] Clear open note when moving back to All Notes
 - [x] Can't return to edit mode when in preview but Markdown disabled remotely
 - [x] Prevent closing app/tab until changes sync'd (@dmsnell)
 - [x] Select first note after first load
 - [ ] Tag Doctor: Maintain tag consistency
	 - [x] Trash tag: remove from all notes
	 - [x] Rename tag: rename in affected notes
 - [x] Add checklists
 - [x] Keyboard shortcuts
	 - [x] Checklist
 - [x] Analytics (@codebykat) https://github.com/Automattic/simplenote-electron/pull/2223
 - [ ] Add UI to show `null` state for analytics toggle (@belcherj @codebykat)
 - [ ] ~Read old persisted in-queue edits (@dmsnell)~ (leaving out for now to avoid the complexity and risk of propagating note corruption)
 - [x] Fix double-note-forwarding on REMOTE_DELETE_FOREVER (local/remote changes)
 - [x] Empty Trash (@codebykat) https://github.com/Automattic/simplenote-electron/pull/2209
 - [x] Highlight search results in note list previews (@belcherj @codebykat)
 - [x] Consider adding back sync indicator in the side bar
 - [ ] Add Electron confirmation dialog when closing with unconfirmed changes
 - [x] Lazy load audit: where can we asynchronously load to cut boot time
 - [x] In history viewer base preview rendering off of whether Markdown preview is selected (currently only shows Markdown preview if note has Markdown enabled) (@codebykat) #2281 and #2282

## Offline
 - [x] Stop writing old revisions or missing revisions over valid ones
 - [x] Separate persistence that can happen idempotently from those that much occur atomically (performance)
    - `noteRevisions` extracted

## Editor
 - [x] Checkbox detection
 - [x] Indenting/outdenting
 - [x] Continue lists on enter (@dmsnell @codebykat)
 - [x] Checkbox toggling
 - [x] Make scrollbars work properly ❓
    - [x] Push scrollbar on narrow view to the edge of the window. It currently floats on the edge of the `<textarea>`
 - [x] Fix editor size ❓
 - [x] Cursor shape over todo item should be pointer
 - [x] Scrollbar slider should look more native
 - [x] Highlight search results (@dmsnell @codebykat @belcherj)
 - [x] Fix selection wonkiness when inserting/removing tasks (@dmsnell)
 - [x] Embed custom font with proper SVGs
 - [x] Prevent the title formatting from following edits to the first line
 - [ ] Firefox inserts two emoji from picker ([Reported to Monaco](https://github.com/microsoft/monaco-editor/issues/2064))
 - [ ] Emoji with visual modifiers not showing visually in editor (Firefox)
 - [x] Insert check list is buggy (add a checklist item give it some text, go to next line, add another, it deletes original)

## Preview

 - [x] Open links properly
 - [x] Copy all HTML if nothing selected #2184
 - [x] Allow for non-markdown preview in revision selector

## Simperium

 - [x] Add client
 - [x] Load notes
 - [x] Build sync meta state
 - [x] Add sync status
 - [x] Update tags when confirming new note
 - [x] Properly delineate remote vs. local edits on notes (`bucket.on` vs. `bucket.channel.on`)
 - [ ] Add note-remove queue vs. directly calling `noteBucket.remove()`
 - [ ] Add tag-remove queue vs. directly calling `tagBucket.remove()`
 - [ ] Schedule routine `index` operations

## Simperium Middleware Operation

 - [x] Don't send any updates during an index operation
 - [ ] Always track last base version for a note
 - [x] Store all revisions for a note, start loading when opening a note
 - [x] Track connection status: red / yellow / green meaning:
	 - "connection is down"
	 - "connected still but haven't heard a heartbeat lately"
	 - "connection is confirmed active"

## Bugs
- [x] Collaboration emails show in note list
- [x] Fix note last updated (JB saw: Note last updated: NaNs ago (Invalid Date))
- [x] Fix styling on text not being vertically centered due to: 'Simplenote Tasks' (Tag pills, log in fields, etc.)
- [x] Could not log out in Electron with unsynced changes

## Enhancements

 - [x] Stylize/move/finish the "last sync'd" time for notes. (@SylvesterWilmott @codebykat @belcherj)
 - [x] Style Server connection in side bar
 - [ ] Refine the notification system
      - [x] Remove notifications code when not available (IE)
      - [x] Make sure we don't spam permissions requests
 - [ ] Update to CSS variables
 - [ ] Stop spamming note-list redraws when updating many notes sequentially
 - [x] Consider adding slight delay when showing the Note not synced yellow triangle
 - [x] Update copy on Notes not synced dialog
 - [ ] When renaming into an existing tag, also rename other tag into which we're renaming (rename `tag1` to `Tag2` should also rename `tag2` to `Tag2`)

dmsnell committed:
Rewrite app core to remove data races and fix data flow
Trap account name when connecting
Type Simperium for real
Re-connect note editing to Simperium; bounce updated notes
Add sync indicators on note-level
remove pending ccid list; small cleanup
Fix or skip or delete failing tests
Bow to the linting gatekeepers
Bow to the linting gatekeepers
Try: remove arm builds to get electron-builder to go
Move all checkbox translation into the editor component
Adjust editor cursor/selection in response to remote updates
Fix edge cases with tasks and selection tracking
Track editor selection direction
Earmark selection tracking on undo/redo for later
Debounce search when receiving updates
Select first note that loads in app
Use search sort index instead of sorting on every search
Fix performance and display issues with note list previews
Allow syncing new notes
Add revision panel back in
Start making tag list sortable and editable
Inline checkbox/task font

I tried using a file-loader and things got complicated with Monaco's fonts
By inlining the font as a data-uri there's no complexity other than maintaining
the font, but it's trivial enough to recreate the font and convert it into
base64 that I felt it was sufficient to let this be.

Continue restoring functionality to tag list
Add some types to tag list interface
Connect to remote tag updates

Small cluterr
Start tracking tag changes
More tag tracking
Add back has-loaded
Allow tag reordering
Allow deleting tags
Fix title formatting/decorations
Remove syntax highlighting and set auto-indent
Support renaming tags; small font fix and type error fix
Allow escaping from preview mode if note un-markdownd
Fix revisions display issue
Reconnect settings to Electron
Add back "insert task" via hotkey
Extract generic bucket providers; sync tag renames
Synchronize preferences
Converge to a single ghost provider
Remove old noteChangeVersion reducer
Start introducing offline persistence
Load and persist revisions separately from the state
Disable revisions button if no revisions are avialable
Fix typo
Defer loading editor and fetching revisions for performance while switching notes
Fix Safari matchMedia bug

Fix Safari Monaco issue: advanced text-wrap crashes

Degrade gracefully in the absence of indexedDB (private mode)

Remove constraint on having indexedDB in platform support

Doubt relevance of test

Refactor code to avoid lint issue

Use published @types/simperium

Sync changes with bucket-queue worker

Use queue function not setTimeout for note sync

Create Note Doctor to find discrepancies with note sync

Warn when logging out with unconfirmed changes

Reconnect the rest of the menu options/settings

Fix Export (#2180)

Fix export on rewrite/no-races

Send notification on note update

Send message when note trashed

Limit notification length and provide default text

Allow logging out when no changes are unconfirmed

Move notifications code into new module

Separate local confirmations from remote updates

(dev) expose buckets to console

Add (un)publish notifications; skip some notification spam

restore tag icons, add overflow for scrolling (#2183)

Fix/search case sensitivity (#2182)

* search should be case insensitive

* fix missing search results due to column alignment

* lowercase term instead of regex

Rewrite tags

Confirm new tags on note creation; delete old tag on rename

Replace Monaco with textarea

Fix: Copy all html if nothing selected (#2184)

* first hack at this, broken

* pass state to useEffect and rename it

* this div is still not working and i am mad

* hook to promise chain, realize clipboard things are hard

* make it work in Safari too

Maintain consistency of login sessions across browser sessions

Fix notifications permissions requests: stop spamming

Fix tag-suggestion tests

Fix lint issues

More lint issues

Add beta warning

Allow exporiting unsyced notes on logout (Electron)

Add logout confirmation dialog

Warn before closing tab with synced changes

Marginally improve cursor when entering task item from hotkey

Stop resizing content editor

Linting issues

Use slightly-faster checkbox replacement

Style last synced indicator (#2203)

Update note list note status icons (#2204)

Add safety for broken/missing indexedDB props: @codebykat

Empty trash (#2209)

* got the basics, not working

* typos in comments

* make it go

Fix tag list CSS (#2210)

* move scrolling to child list

* add some right padding to avoid overlapping with scrollbar

* more small fixes

Reset some optimizations in the editor

Removes bottom border in the tag list (#2214)

Updates CSS for server connection (#2213)

Fix scroll bars on narrow line length (#2211)

When the narrow line length is set the scroll bar was not on the edge of the screen. This moves it back to the edge.

Update/offline badge (#2202)

Moves the offline badge from the bottom right to the note toolbar.

Filter collaboration (email) tags from tag list (#2205)

Updates unsynced notes log out dialog (#2206)

Display unknown in last updated when it is unknown (#2212)

Display unknown in the last synced indicator when we do not know when it was last synced.

Move all tasklist insertion into the editor itself

Use function-ref to boot editor; update text transform in reducer

Insert checkboxes through editor

Continue lists

Fix interaction of lists and tasks

Fix font issues

Add indent/outdent support

Stop indenting lines that follow blank lines

Show checkboxes in previews

Preserve some markdown settings

[rewrite branch] Change the unsynced notes icon to a less-scary one (#2218)

This changes the icon when a note has unsynced changes to the cloud sync icon.

Also changes it to blue and changes both this and the pin to simplenote-blue-20 in dark mode.

Fixes missing ids in note export (#2220)

Clear note list height cache when window width changes

Stop crashing when viewing a trashed note

Remove all text processing

make condensed actually condense (#2222)

[rewrite branch] update note list status icons (#2221)

* add smaller icons and restyle

* forgot to commit the style fix for the bottom border

* larger icons more padding

* clean up styles

Bring back Monaco

Spin sync indicator

only spin the sync icon, not the publish icon

Stop sync spinner when offline

Only load window-close-confirmation to browser

Update the editor width css to allow narrow/wide views (#2224)

Remove the add/confirm step which isn't necessary

Remove confirm step for tags

Fixes styling of offline badge (#2234)

Updates the color of the link in the logout confirm dialog (#2235)

Dont scroll editor past last line (#2238)

When you scroll a note it currently scrolls the note completely off the screen. This PR changes it so it does not scroll past the end of the note. I like the current behavior and think we should add this back as an option.

Allow scroll bar to be any height fixing it only scrolling partially to the bottom (#2239)

We had a fixed height on the scroll bar. This caused it to look like it hadn't scrolled all the way to the bottom. This removes that max height so that the scroll bar displays correctly.

Fixes editor width when narrow line length is selected (#2237)

My previous fix for the editor width only worked on the left margin. This fixes it so the editor width is constrained on both sides. It does however cause the scroll bar to be just inside the narrow view and not on the edge of the screen, see screenshot. I am not sure how to fix it to push it to the side.

This PR also fixes the jerkiness of the note when it opens. The padding applied to the content before the editor is ready is now the same as it is once Monaco renders.

Update height of note list item when condensed (#2246)

Clean up disable restore note button when is newset revision (#2248)

The disable on the restore note button when you are currently on the most recent revision is CSS and was not actually disabling the functionality. Additionally, the logic failed when you first opened the revisions panel. This gets everything into tip top.

Use distinct name for indexedDB persistence

Checks for Notification on window (#2225)

* Checks for Notification on window

* Use optional chaioning

* Dynamically load change announcer

Co-authored-by: Dennis Snell <dennis.snell@automattic.com>

Try for some easy wins on optimizing load

Hides last syncd message when it is unknown (#2236)

Hides the last syncd when it is unknown

Autolist behavior

Only use fast edit mode for long notes

Continue task lists too

Allow bullets at start of line, stop logging actions

Remove some pre-prod TODO items

Transform note content properly when clicking on tasks

Stop trying to fully-load an editor when it's gone

Updates to version 2.0.0-beta1 (#2252)

Disable note doctor for now; uncover bugs easier

Refactor some tag rename logic

[rewrite branch] Add analytics as middleware (#2223)

* centralize analytics again
 - not working attempt to convert to meta key
 - fixed addCollaborator args
 - add removeCollaborator reducer with withEvent
 - Add COLLABORATOR actions
 - remove invalid comments
 - remove unused setAccountName and some more comments
 - trying event queue to add application_opened before preferences are loaded
 - documented bugs :[

* Update analytics setting when bucket connects

* Set initial accountName in Redux init vs. dispatched action

* Send index request for preferences

* Persist account name properly

Co-authored-by: Dennis Snell <dennis.snell@automattic.com>

Remove tri-state analytics toggle

Sequence analytics updates

Fix analytics during import: use Redux action not direct function call

Revert "Add beta invitation banner (#2208)"

This reverts commit 2ea1d3da9457f58949aa4700905b03af33100d3b.

Take that out of there…

Fixes note underline in firefox (#2261)

Fixes the underline on detected links in Firefox.

Fixes the polyfill for Promise to make IE work (#2260)

The polyfils need to be moved up to boot so that they exist everywhere for Internet Explorer. We should look later at removing them when not necessary in browsers.

Stop removing Electron event listeners after adding the new ones

Add/sentry (#2262)

Adds Sentry to app so that we can track and fix errors

Harmonize menu bar and settings in sidebar (#2263)

The web app has settings/keyboard shortcuts/about items in the navigation bar but those are absent from Electron.
Electron already provides access to the settings through the OS menu but it's confusing to jump between the web and
the electron app to discover those items are missing in one place.

In this PR we stop hiding the settings and other items in the navigation sidebar when running in Electron.

The Electron OS menu has also been missing "Empty Trash" and "Sign Out" as the native apps have. "Sign Out" is particularly
convenient to have there and its absence makes logging-out particularly inconvenient.

As a result of the overall state refactor in the app we're adding those menu items to make it more convenient again.

Sync notes affected by a tag rename

Only select next note when trashing/deleting/restoring if opened note is affected note

Continue all tasks as unchecked

Remove Simplenote font from body fixing line height (#2274)

Support inter-note linking with Markdown syntax

Prefil full link syntax with note title when copying inter-note link

Drop lonely bullets on newline

create custom themes for Monaco (#2272)

[Rewrite Branch] Fixes undo redo selectall (#2181)

This fixes the undo, redo, and selectall in the Electron menu

add Tasks font back to preview wrapper (#2282)

[rewrite branch] Fix plain-text revision preview (#2281)

* always show plain-text note if we're showing revisions

* actually let's look at editMode so we can still preview rendered content

* move logic to mapStateToProps

disable bracket highlighting (#2283)

Fix task font by adding descent adjustment

Use wifi signal icons for server connection

[Rewrite Branch] Fix printing of notes (#2271)

This PR adds a display none div that contains the content of a note. When the user prints it displays that div and hides other content on the page.

Adding a display none div with the content allows us to do this purely with CSS rather than watching for print events and potentially running into race conditions with React rendering.

Clear React warnings in un-editable input components

(cherry picked from commit 5dd1cfbad1db305f6828193e0de5ec794d67e6a8)

Fix preview note view: grab proper note object and update with props

In #2281 we fixed the note preview so that it would only render a
Markdown-enabled note if the preview mode in the editor is set.

Unfortunately we missed a dependency in the `useEffect()` callback
and so the preview would only render the corresponding view to the
settings passed when it was first opened. This has been fixed by
adding the proper dependency.

Additionally we were only getting the Markdown setting for the passed
note, such as when viewing revisions from a note's history. For normal
preview views we missed examining the opened note. We have fixed this
by picking the appropriate note object to inspect when determining
which mode to render.

Remove test-account pre-filler: imlpement better later on

Stop updating note cell if it's no longer mounted

Reflow editor layout when changing line length

Pulled from #2247, props to @belcherj for the work

Delay re-running the editor resize to fix text-clipping after line-width change

Use updated connection icons: thanks @SylvesterWilmott!

Use updated sync icon so it spins clockwise

[rewrite branch] Fix keyboard shortcuts (#2284)

* disable keybindings Monaco is overriding

* also remove selectHighlights

* remove cursorUndo, jumpToBracket

* first hack at checkbox insertion

* change tag toggle back to a toggle

* Expand detection of tasks and bullets for various line contents

* fix cmd+A select all

* fix tag/editor toggle

* use a context key to always show keyboard hint in the context menu

* use editorCommand instead of appCommand for insertChecklist

* fix Electron menu items

Co-authored-by: Dennis Snell <dennis.snell@automattic.com>

[Rewrite Branch] Add debounce of 2 seconds to analytics for note edit (#2289)

* Add debounce of 2 seconds to analytics for note edit

* minor cleanup

Co-authored-by: Kat Hagan <kat@codebykat.com>

Audit note updates: update modification stamp; re-index note on changes

check for editor focus and fallback to native functionality (#2293)

Add missing style to fix react-virtualized autoSizing (#2294)

Adds highlighting in note for search query (#2292)

* Add document highlight provider to highglight search results

In this patch we're adding a new document highlight provider to highlight the
search results in a note. This may be a useful approach because it relies
entirely on Monaco's API to provide the functionality. Unfortunately in my
exploration it wasn't clear to me to how get this to highlight when the cursor
isn't on any word. The API appears to be designed to highlight related words to
the word under the cursor. If the cursor is on an empty line then nothing else
gets highlighted and the provider never gets called.

If we can force-trigger the highlighting to run this might be workable. We need
to do that and also be able to select an active highlight if we want to use
this for search.

We might have `editor.action.wordHighlight.trigger` as an action to call but I
gave up trying to figure out how to properly call it.

* Change highlight colors

* Use decorations

* Fix css class

* Remove uneeded highlighting styles

* Combine title and search decorators

* Make occurrencesHighlight false again

* Account for notes with no content

* Make highlight text white when in light mode

* Cleanup from code review

* Gate this to monaco so it doesnt run on fast view

Co-authored-by: Dennis Snell <dennis.snell@automattic.com>

Fixes line wrap in markdown view (#2296)

When you have a really really long word it can cause a horizontal scroll in markdown view. This will make words wrap.

Fixes the width of content in the editor (#2297)

Notes have a maximum width which is incorrect. This makes the max column width 400. Using the bounded strategy it will fill viewport till it hits 400 columns. By using advanced it fills in available space better.

[rewrite branch] Too many scrollbars (#2299)

I was testing on Windows/Edge and noticed we've got some scrollbars on note-detail and friends that are cluttering up the place.

The Monaco container is inside of .note-detail, and handles its own overflows and scrolling, so we don't need these.

Special-cased the Markdown preview; make sure you can scroll in preview if necessary!

[Rewrite Branch] Use Monaco context menu for editor actions (#2280)

Moves to the Monaco context menu. This enables undo/redo/select all/copy/cut from the context menu.

Disable Beta banner

Remove code that is causing continuous indexing events (#2304)

[rewrite branch] Fix various CSS issues (scrollbars, icons, font sizes) (#2303)

* tag list scrollbar auto

* keybindings scroll auto

* fix huge menu icon

* hide horizontal scroll on note detail

* first line should be relative correctly

* fix note list scrolling

Only request revisions when the revisions panel is opened (#2305)

Change how revisions are requested to not spam Simperium on every note switch.

only dispatch SET_ANALYTICS if the value is different from what we have in state (#2306)

Revert "only dispatch SET_ANALYTICS if the value is different from what we have in state (#2306)" (#2307)

This reverts commit e6ab7116c645db527ed05b291140d8ea3e0f654d.

[rewrite branch] Only dispatch SET_ANALYTICS if it has changed, take 2 (#2308)

* check state not bucket

* remove random import

* coerce value

Relax preferences updates

Remove uneccesary divs (#2314)

[rewrite branch] Logout confirmation dialog: use color studio for all hex colors, fix typo (#2315)

* replace hex codes with color studio

* typo fix

Fixes the css so all pages print (#2319)

There were three issues with printing. The position relative shows more than the first page. The overflow auto allows all the pages to print. I have a note that should bee 22 pages but only 11 were printed without the overflow auto. The last issue is that display flex in Firefox causes only the first page to print. So added display: block to fix that.

Add persistent "preferences" bucket

Fixes the editor height in Safari (#2320)

Fixes safari height and Chrome editor width (#2321)

add a wrapper span so we can pass a disabled button (#2322)

Temporarily disconnect preferences bucket

(cherry picked from commit 5e954bb9c2cebc58d09b984df97f1b6c76fc51ac)

move span inside tooltip (#2324)

Dont find highlights if no search (#2323)

[Rewrite Branch] Adds the new search interface (#2313)

* Adds the new search interface

* Fix logic when note changes

* Revert selection highlight color

* Formalize logic for content and reduce

Co-authored-by: Dennis Snell <dennis.snell@automattic.com>

Version bump

Create persistent preferences bucket, adding logic fixes

Right align rtl languages (#2332)

When a line is a rtl language such as Arabic it should be aligned to the right. This PR floats it to the right.

Bump version to rc3

[rewrite branch] Align the scrollbar to the right (#2298)

* hack some padding

use normal left padding

remove decoration margin

special case markdown preview

refactor css

fix scrollbar highlights

moved this css change elsewhere

try this.contentDiv

move createDiv ref

* use decorator for padding, remove title size

* don't check checkboxes from the gutter

* helps to actually commit changes after rebasing

* refactor some stuff, thanks Dennis :)

fix editor padding with wide screen and gofast (#2333)

[staging] Fix keyboard shortcuts on Edge and non-English layouts (#2334)

* change all codes to keys

* lowercase key because Edge

fix search selection issues (#2335)

* fix search selection issues

* don't focus the editor on load

Various linting fixes (#2337)

This PR cleans up some lint warnings that have been introduced over time:

- Remove a console.log where it seemed unnecessary
- Changed a console.log to debug where it made sense to keep the logging
- Changed some variable names to distinguish from variables declared in an upper scope

Updates styling for rtl lanaguage lines (#2338)

Adds correct spacing to the right side of the editor for RTL languages.

Delete unused variables (#2329)

These two sass variables are unused. This removes them as they are not necessary.

Move last-sync indicator into note info panel

Bump release-candidate version

Small refactor

belcherj commented:
Tested: Test text file import

I imported several text file notes of varying complexity on staging.simplenote.com

belcherj commented:
What is wrong with the scroll bar? I think it works much better than previously. I like being able to scroll the text all the way up so I have a clean slate. I think this is also a feature that has been requested a lot. @SylvesterWilmott what you think?

belcherj commented:
- [ ] Capital letters in search cause search to fail
- [ ] Export notes doesn't work in electron
- [ ] Spellchecker in electron is broken 

Electron Menu:

- [ ] Undo
- [ ] Redo
- [ ] Select all
- [ ] Sort Type
- [ ] Line length
- [ ] Tags, sort alphabetically
- [ ] Theme did not match selected theme in menu initially
- [ ] Focus mode
- [ ] Insert Checklist is buggy

codebykat commented:
Impossible to use in Safari, this modal comes up and immediately comes back regardless of what you click. You can't even quit the app normally 🙃 

<img width="441" alt="Screen Shot 2020-07-01 at 11 37 58 AM" src="https://user-images.githubusercontent.com/52152/86279827-8425ab00-bb8f-11ea-9d18-8eb79c148d8a.png">

SylvesterWilmott commented:
>What is wrong with the scroll bar? I think it works much better than previously. I like being able to scroll the text all the way up so I have a clean slate.

This is something I have seen in a few editors, however, I would recommend only providing a 25% or 33% space below the note to type and also (if possible) not allowing this behaviour unless the note is long enough to be scrollable. Notion uses a similar logic.

belcherj commented:
- [ ] Fix note last updated (Note last updated: NaNs ago (Invalid Date))
- [ ] Style/Move Note last updated
- [ ]  Fix styling on text not being vertically centered due to: 'Simplenote Tasks' (Tag pills, log in fields, etc.)
- [ ] Add a 2 second delay showing yellow warning not-syncd
- [ ] Server connection, make it look like the old one and put server connection in hover window
- [ ] Stop showing emails in the note list

belcherj commented:
Things I found while testing:

- [ ] Jump to next match in note and jump to the previous match in note not implemented
- [ ] Insert checklist hotkey does not function
- [ ] Switching from narrow line width to full needs to refresh editor
- [x] Restore note button is disabled on the latest change but only after looking at other revisions
- [x] Condensed note list looks a bit off, see image

<img width="329" alt="Screen Shot 2020-07-28 at 12 49 22 PM" src="https://user-images.githubusercontent.com/6817400/88696104-d1cefe00-d0d0-11ea-8fd8-8bc27be590a4.png">

Will work these into the PR description.

dmsnell commented:
Rebase diffs are at https://gist.github.com/dmsnell/d98b5b3dc825d375528fe087fe297419

The `in-dev` diff shows the changes from the previous merge base to the `development` HEAD
```bash
git diff 9a0d3bf99a2f1bdb8e8179946833cfba21df15ae 2ea1d3da9457f58949aa4700905b03af33100d3b
```

The `in-branch` diff shows the changes from the previous branch tip to the new branch tip
```bash
git diff 19a283a96d4262617275dda7c5dc3cc2812a7c70 b359787857ed49f644c5f91e1f98fbdd09d4fbef
```

Assuming I rebased properly these two patches should be more or less equal with each other. To wit there are a few changes as a result of the fact that the rewrite branch changes certain things, such as `exportZipArchive` to `exportNotes`.

The third file in [the gist](https://gist.github.com/dmsnell/d98b5b3dc825d375528fe087fe297419) therefore contains the diff of the two diffs. There's a decent amount of noise in there due to commit hashes being different but that should compare the list of changes the happened in `develop` after the previous fork of the rewrite branch and the list of changes made in order to move the rewrite branch to the new base.

**Revisions:**
 - Updated after performing the rebase again one commit at a time.

belcherj commented:
Things found in testing Safari, latest, OSX:

- [ ] When permanently deleting notes in one browser updates the selected note in another browser when all notes is selected.
- [ ] Create checkbox, mark as completed, enter down, checkbox created is marked as complete.
- [ ] Buttons in revision history and trash view do not have a proper line-height
- [ ] Line height for search highlights is incorrect.
- [ ] Text on staging is #d4d4d4 not #fff in dark mode (#2265, fix in #2272)
- [ ] Tag names in edit mode have bottom chopped off, edit tag a `qqq`, observe
- [ ] In search field descenders are cut off, type `qqq`
- [ ] When coming back online sever connection does not  go fully green
- [ ] Printing in edit mode results in blank page, preview works

<img width="194" alt="Screen Shot 2020-08-10 at 3 56 24 PM" src="https://user-images.githubusercontent.com/6817400/89825259-50d92300-db22-11ea-96d9-dd5be58154f1.png">

belcherj commented:
Things found in testing Firefox, latest, OSX:

- [ ] When you have the publish dialog open in Firefox and Chrome, toggle on publishing, it then toggles off
- [ ] Managed to get a note to go out sync after being offline
- [ ] Select all in context menu doesn't work

codebykat commented:
We need to fix analytics performance as well (see https://github.com/Automattic/simplenote-electron/issues/2172#issuecomment-670800192 )

codebykat commented:
## Results of keyboard shortcut testing

Edit to update! These have mostly been fixed in #2284 

I tested all our app shortcuts (the ones in the dialog) in OSX/Firefox and Electron.

### Monaco is overriding the following shortcuts (they do not work when editor has focus):
- [x] `cmd` + `/` (on OSX it's only catching `cmd`, `ctrl` works)
- [x] `ctrl` + `shift` + `C` -- insert checklist item
- [x] `cmd` + `L` to jump to URL in browser, and probably a lot more -- we should disable most if not all of the Monaco shortcuts (see #2279 )

### Totally nonfunctional (does not work no matter where focus is):
- [x] `ctrl` + `shift` + `Y` -- toggle editing note/tags

### Misc issues
- [x] toggle tag list (`ctrl` + `shift` + `U`) only OPENS the tag list, does not close it
- [x] `ctrl` + `shift` + `L` (toggle note list (on narrow screens)) -- works but on full-width it opens a Monaco popup

### Search shortcuts don't work because we don't have matches within note. On Firefox, `ctrl` + `F` triggers in-editor search with missing fonts.
- [ ] `ctrl` + `F` // should show browser search but does not. Shows Monaco search in Firefox, nothing in Electron
- [ ] `ctrl` + `G` // jump to next match
- [ ] `ctrl` + `shift` + `G` // jump to previous match
  
<img width="485" alt="Screen Shot 2020-08-18 at 2 56 27 PM" src="https://user-images.githubusercontent.com/52152/90570348-56f38300-e164-11ea-9be5-c4eff063656e.png">

codebykat commented:
I added search as a blocker. Here's my current list of known bugs:

* need to implement search #2313; issues with search PR noted [here](https://github.com/Automattic/simplenote-electron/pull/2313#issuecomment-685326207)
* note with long lines (not just the title, any line) allows editor to horizontally scroll, content gets hidden
* issues with printing: https://github.com/Automattic/simplenote-electron/issues/2317
   * Markdown preview only prints one page
   * editor printing seems to mess up the formatting of the last page or couple pages
* Scrollbar alignment: [in progress PR, needs an update but should be feasible] https://github.com/Automattic/simplenote-electron/pull/2298

codebykat commented:
**Current list of known issues:**

* Some search follow-ups to be addressed (these are mostly styling; see #2331)
* RTL language support is not great (selection is wonky, scrollbar overlaps)
* Firefox emoji picker and accents insert double characters (#2327, https://github.com/microsoft/vscode/issues/106392)
* Doesn't work well in mobile browsers (double scrollbars, context menu not working, probably other issues as well)
* Deleted notes sometimes come back

**Unverified/unreproduced:**

* Some issues reported with zoom in Chrome
* Printing weirdness (maybe fixed, I haven't seen it lately) https://github.com/Automattic/simplenote-electron/issues/2317
* EOL character weirdness https://github.com/Automattic/simplenote-electron/issues/2309

dmsnell commented:
Squashed and force-pushed from 667b643 to 957a513

dmsnell commented:
Force-pushed from 957a513 to ca8d3ae
